From 5e68e994897d006a22aa927ce23bd34abb72ef96 Mon Sep 17 00:00:00 2001
From: Lennart Poettering <lennart@poettering.net>
Date: Thu, 12 Jan 2012 05:09:24 +0100
Subject: [PATCH 206/217] journal: if the syslog forwarder socket is full,
 then don't block (cherry picked from commit
 7c8bbccd071a9a12efdd8a7770b53d7786eb3a9c)

---
 src/journal/journald.c |    8 ++++++++
 1 files changed, 8 insertions(+), 0 deletions(-)

diff --git a/src/journal/journald.c b/src/journal/journald.c
index d020309..87f1e5b 100644
--- a/src/journal/journald.c
+++ b/src/journal/journald.c
@@ -706,6 +706,11 @@ static void forward_syslog_iovec(Server *s, const struct iovec *iovec, unsigned
         if (sendmsg(s->syslog_fd, &msghdr, MSG_NOSIGNAL) >= 0)
                 return;
 
+        /* The socket is full? I guess the syslog implementation is
+         * too slow, and we shouldn't wait for that... */
+        if (errno == EAGAIN)
+                return;
+
         if (ucred && errno == ESRCH) {
                 struct ucred u;
 
@@ -719,6 +724,9 @@ static void forward_syslog_iovec(Server *s, const struct iovec *iovec, unsigned
 
                 if (sendmsg(s->syslog_fd, &msghdr, MSG_NOSIGNAL) >= 0)
                         return;
+
+                if (errno == EAGAIN)
+                        return;
         }
 
         log_debug("Failed to forward syslog message: %m");
-- 
1.7.8.3

